Load shedding

Ladicek opened the pull request:
Related to #36543

Ladicek committed:
Load Shedding: TCP Vegas, request priority and request cohort

The overload detector uses a TCP Vegas based algorithm,
as implemented by Netflix Concurrency Limiters.

Priority load shedding uses 5 priority levels and 128 cohorts.
A simple cubic function is used to determine the threshold that
current CPU load has to reach to reject the current request.

Ladicek commented:
Draft because this is very much work in progress. Sharing to maybe get some initial feedback.

If this is considered to be too niche for core Quarkus, I'd be fine with moving to Quarkiverse.

CC @ahus1 @vmuzikar @tmonney

cescoffier reviewed (commented):
Looks great!

First, I like the overload algorithm you are using. Any reference on it? 
Then, we have a connection limiter in Quarkus already, should we deprecate this in favor of the load shedding?

Also, I think it needs to be tested with long-running connection (gRPC streams, SSE, web sockets). 

I'm also thinking that on overload, we may need to adjust the readiness Kube probe. WDYT?

Ladicek commented:
The algorithm is a simplified version of https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/VegasLimit.java, I think there's a link in the javadoc already.

> Then, we have a connection limiter in Quarkus already, should we deprecate this in favor of the load shedding?

Ah do we? I had no idea. Where can I learn more?

> I think it needs to be tested with long-running connection (gRPC streams, SSE, web sockets).

Good point, I didn't try that at all. I'll check, but I doubt I'll see anything meaningful. The present implementation is heavily oriented to request/response style of interaction. Occasional stream likely won't do anything, and a streaming-heavy application would require a more involved implementation.

> I'm also thinking that on overload, we may need to adjust the readiness Kube probe. WDYT?

Hmm, I'm not sure about that. That would be super coarse-grained.

Ladicek commented:
Just marked as ready for review. I fixed a couple of bugs in the implementation and added some rudimentary documentation.

The feature is not very heavily tested. @franz1981 do you think it would be possible to test this in a perf lab? I don't really know how that works, but I guess we have a test that stresses Quarkus application above its capacity, where this would help?

Ladicek commented:
Rebased and fixed the conflict.

Ladicek commented:
Rebased and fixed a few tiny issues. I believe this is ready now.

cescoffier reviewed (approved):
I think we should merge it and iterate.

gsmet commented on extensions/load-shedding/runtime/pom.xml:
```diff
@@ -0,0 +1,44 @@
+
+    <artifactId>quarkus-load-shedding</artifactId>
+
+    <name>Quarkus - Load Shedding - Runtime</name>
```
I think we need a description here.

Ladicek replied:
Will add.

Ladicek replied:
Done.

gsmet commented on extensions/load-shedding/runtime/src/main/resources/META-INF/quarkus-extension.yaml:
```diff
@@ -0,0 +1,13 @@
+  keywords:
+  - "fault-tolerance"
+  - "load-shedding"
+  guide: "https://quarkus.io/guides/load-shedding"
```
isn't it `load-shedding-reference`?

Ladicek replied:
Ouch! Good point, will change.

Ladicek replied:
Done.

gsmet reviewed (commented):
I added two small comments. I don't mind you merging and addressing them later so that we avoid some more conflicts.
